### PR TITLE
[Build] use pigz to speed up a build

### PR DESCRIPTION
**Pull request by kv-y**
#### Why I did it

It's possible to speed up some parts of a build using parallel compression/decompression.
This is especially important for build_debian.sh.

#### How I did it

pigz is a parallel implementation of gzip: https://zlib.net/pigz/
Some programs like docker and mkinitramfs can automatically detect and use it instead of gzip.
For tar we need to select it directly.

To enable this feature you need to set GZ_COMPRESS_PROGRAM=pigz

### CODE REVIEW

**Comment by kv-y** [comment]
Comparison. 
Old:
```
01:45:17 [ building ] [ target/sonic-broadcom.bin ] 
02:17:10 [ finished ] [ target/sonic-broadcom.bin ]
32 min 53 sec

02:47:54 [ building ] [ target/sonic-mellanox.bin ] 
03:02:48 [ finished ] [ target/sonic-mellanox.bin ] 
14 min 54 sec

01:25:14 [ building ] [ target/sonic-marvell-armhf.bin ] 
01:42:54 [ finished ] [ target/sonic-marvell-armhf.bin ] 
17 min 40 sec

01:01:55 [ building ] [ target/sonic-vs.img.gz ] 
01:34:20 [ finished ] [ target/sonic-vs.img.gz ] 
32 min 25 sec
```

new
```
02:20:48 [ building ] [ target/sonic-broadcom.bin ] 
02:44:59 [ finished ] [ target/sonic-broadcom.bin ] 
24 min 11 sec

01:52:04 [ building ] [ target/sonic-mellanox.bin ] 
02:04:06 [ finished ] [ target/sonic-mellanox.bin ]
12 min 2 sec

01:12:03 [ building ] [ target/sonic-marvell-armhf.bin ] 
01:31:48 [ finished ] [ target/sonic-marvell-armhf.bin ] 
19 min 45 sec

01:31:40 [ building ] [ target/sonic-vs.img.gz ] 
01:55:14 [ finished ] [ target/sonic-vs.img.gz ]
23 min 34 sec
```

It may depends from different factors, but I got 15-25% improvements in build time of final debian image (build_debian.sh) with pigz.  Most noticeable changes here: `tar -I pigz -cf platform.tar.gz -C $PLATFORM_DIR`

File: build_debian.sh [hunk 1]
```diff
@@ -141,7 +141,7 @@ fi
 ## 2. mount supports squashfs
 ## However, 'dpkg -i' plus 'apt-get install -f' will ignore the recommended dependency. So
 ## we install busybox explicitly
-sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install busybox linux-base
+sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install busybox linux-base pigz
```
Comment (lguohan) [inline build_debian.sh:144]: do we need to install pigz in the base image?

**Comment by lguohan** [comment]
can you make it optional so that people can try it out before we make it default?

File: build_debian.sh (hunk 1 above)
Comment (lguohan) [reply build_debian.sh:144]: make sense. can you make this pigz as optional as suggested before?

**Comment by kv-y** [comment]
> can you make it optional so that people can try it out before we make it default?

I think it's possible. I'll update this PR later then.

**Comment by Kalimuthu-Velappan** [comment]
> Comparison. Old:
> 
> ```
> 01:45:17 [ building ] [ target/sonic-broadcom.bin ] 
> 02:17:10 [ finished ] [ target/sonic-broadcom.bin ]
> 32 min 53 sec
> 
> 02:47:54 [ building ] [ target/sonic-mellanox.bin ] 
> 03:02:48 [ finished ] [ target/sonic-mellanox.bin ] 
> 14 min 54 sec
> 
> 01:12:03 [ building ] [ target/sonic-marvell-armhf.bin ] 
> 01:31:48 [ finished ] [ target/sonic-marvell-armhf.bin ] 
> 19 min 45 sec
> 
> 01:01:55 [ building ] [ target/sonic-vs.img.gz ] 
> 01:34:20 [ finished ] [ target/sonic-vs.img.gz ] 
> 32 min 25 sec
> ```
> 
> new
> 
> ```
> 02:20:48 [ building ] [ target/sonic-broadcom.bin ] 
> 02:44:59 [ finished ] [ target/sonic-broadcom.bin ] 
> 24 min 11 sec
> 
> 01:52:04 [ building ] [ target/sonic-mellanox.bin ] 
> 02:04:06 [ finished ] [ target/sonic-mellanox.bin ]
> 12 min 2 sec
> 
> 01:25:14 [ building ] [ target/sonic-marvell-armhf.bin ] 
> 01:42:54 [ finished ] [ target/sonic-marvell-armhf.bin ] 
> 17 min 40 sec
> 
> 
> 01:31:40 [ building ] [ target/sonic-vs.img.gz ] 
> 01:55:14 [ finished ] [ target/sonic-vs.img.gz ]
> 23 min 34 sec
> ```
> 
> It may depends from different factors, but I got 15-25% improvements in build time of final debian image (build_debian.sh) with pigz. Most noticeable changes here: `tar -I pigz -cf $OLDPWD/$FILESYSTEM_DOCKERFS -C ${DOCKERFS_PATH}var/lib/docker .`



Thanks for your suggestion. As you said, I also used this in the binary optimization PR. 
https://github.com/sonic-net/sonic-buildimage/pull/10718
I will sync up my changes based on your PR ENV flags.

**Comment by kv-y** [comment]
BTW, there are some places we can also use pigz:
https://github.com/sonic-net/sonic-buildimage/blob/9988ff888b0e797630d9b799129938d61e891c28/scripts/build_debian_base_system.sh#L103

But need to check is it really a bottleneck like compressing dockerfs or not.
I'll probably do a separate PR if found that it's useful.

Also one more interisting fact:
Though  we don't install pigz directly to sonic-slave we already use it indirectly in sonic-slave (e.g. docker load) because it's installed together with docker as suggested package.

```
Step 69/76 : RUN apt-get install -y docker-ce=5:20.10.21~3-0~debian-buster docker-ce-cli=5:20.10.21~3-0~debian-buster
...
The following NEW packages will be installed:
apparmor containerd.io docker-ce docker-ce-cli docker-ce-rootless-extras
docker-scan-plugin iptables libnftables0 nftables pigz
...
http://debian-archive.trafficmanager.net/debian buster/main amd64 pigz amd64 2.4-1 [57.8 kB]
```
https://dev.azure.com/mssonic/be1b070f-be15-4154-aade-b1d3bfb17054/_apis/build/builds/185360/logs/64

**Pull request merged**